<a href="https://colab.research.google.com/github/yush582/genAI/blob/main/Hands_on_Notebook_Pandas%20(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# importing the libraries
import numpy as np
import pandas as pd

### 1.11 & 1.12 Pandas -  Series and DataFrames

**Pandas Series**
* Pandas Series is a one-dimensional labeled array/list capable of holding data of any type (integer, string, float, python objects, etc.).
* The labels are collectively called index.
* Pandas Series can be thought as a single column of an excel spreadsheet and each entry in a series corresponds to an individual row in the spreadsheet.

In [2]:
# creating a list of price of different medicines
med_price_list = [55,25,75,40,90]

# converting the med_price_list to an array
med_price_arr = np.array(med_price_list)

# converting the list and array into a Pandas Series object
series_list = pd.Series(med_price_list)
series_arr = pd.Series(med_price_arr)

# printing the converted series object
print(series_list)
print(series_arr)

0    55
1    25
2    75
3    40
4    90
dtype: int64
0    55
1    25
2    75
3    40
4    90
dtype: int64


* We can see that the list and array have been converted to a Pandas Series object.
* We also see that the series has automatically got index labels. Let's see how these can be modified.

In [3]:
# changing the index of a series
med_price_list_labeled = pd.Series(med_price_list, index = ['Omeprazole','Azithromycin','Metformin','Ibuprofen','Cetirizine'])
print(med_price_list_labeled)

Omeprazole      55
Azithromycin    25
Metformin       75
Ibuprofen       40
Cetirizine      90
dtype: int64


**Performing mathematical operations on Pandas Series**

* The price of each medicine was increased by $2.5. Let's add this to the existing price.

In [5]:
# adding 2.5 to existing prices
med_price_list_labeled_updated = med_price_list_labeled + 2.5
med_price_list_labeled_updated

Omeprazole      57.5
Azithromycin    27.5
Metformin       77.5
Ibuprofen       42.5
Cetirizine      92.5
dtype: float64

* A new price list was released by vendors for each medicine. Let's find the difference between new price and the old price

In [6]:
new_price_list = [77, 45.5, 100, 50, 80]
new_price_list_labeled = pd.Series(new_price_list, index = ['Omeprazole','Azithromycin','Metformin','Ibuprofen','Cetirizine'])
print(new_price_list_labeled)

Omeprazole       77.0
Azithromycin     45.5
Metformin       100.0
Ibuprofen        50.0
Cetirizine       80.0
dtype: float64


In [7]:
print('Difference between new price and old price - ')
print(new_price_list_labeled - med_price_list_labeled_updated)

Difference between new price and old price - 
Omeprazole      19.5
Azithromycin    18.0
Metformin       22.5
Ibuprofen        7.5
Cetirizine     -12.5
dtype: float64


**Pandas DataFrame**

Pandas DataFrame is a two-dimensional tabular data structure with labeled axes (rows and columns).

**Creating a Pandas DataFrame using a list**

In [9]:
student = ['Mary', 'Peter', 'Susan', 'Toby', 'Vishal']
df1 = pd.DataFrame(student,columns=['Student'])
df1

Student
0    Mary
1   Peter
2   Susan
3    Toby
4  Vishal

**Creating a Pandas DataFrame using a dictionary**

In [10]:
# defining another list
grades = ['B-','A+','A-', 'B+', 'C']

# creating the dataframe using a dictionary
df2 = pd.DataFrame({'Student':student,'Grade':grades})
df2

Student Grade
0    Mary    B-
1   Peter    A+
2   Susan    A-
3    Toby    B+
4  Vishal     C

**Creating a Pandas DataFrame using Series**

The data for total energy consumption for the U.S. was collected from 2012 - 2018. Let's see how this data can be presented in form of data frame.

In [11]:
year = pd.Series([2012,2013,2014,2015,2016,2017,2018])
energy_consumption = pd.Series([2152,2196,2217,2194,2172,2180,2258])

df3 = pd.DataFrame({'Year':year,'Energy_Consumption(Mtoe)':energy_consumption})
df3

Year  Energy_Consumption(Mtoe)
0  2012                      2152
1  2013                      2196
2  2014                      2217
3  2015                      2194
4  2016                      2172
5  2017                      2180
6  2018                      2258

**Creating a Pandas DataFrame using random values**

For encryption purposes a web browser company wants to generate random values which have mean equal to 0 and variance equal to 1. They want 5 randomly generated numbers in 2 different trials.

In [12]:
# we can create a new dataframe using random values
df4 = pd.DataFrame(np.random.randn(5,2),columns = ['Trial 1', 'Trial 2'])
df4

Trial 1   Trial 2
0  1.169058 -0.082104
1  0.297347 -0.215480
2  1.287332 -0.133845
3  2.363806 -0.336751
4 -0.845014  0.173023

### 1.13 & 1.14 Pandas - Accessing and Modifying

**Accessing DataFrames**

The data of the customers visiting 24/7 Stores from different locations was collected. The data includes Customer ID, location of store, gender of the customer,  type of product purchased, quantity of products purchased, total bill amount. Let's create the dataset and see how to access different entries of it.

In [13]:
# creating the dataframe using dictionary
store_data = pd.DataFrame({'CustomerID': ['CustID00','CustID01','CustID02','CustID03','CustID04']
                           ,'location': ['Chicago', 'Boston', 'Seattle', 'San Francisco', 'Austin']
                           ,'gender': ['M','M','F','M','F']
                           ,'type': ['Electronics','Food&Beverages','Food&Beverages','Medicine','Beauty']
                           ,'quantity':[1,3,4,2,1],'total_bill':[100,75,125,50,80]})
store_data

CustomerID       location gender            type  quantity  total_bill
0   CustID00        Chicago      M     Electronics         1         100
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50
4   CustID04         Austin      F          Beauty         1          80

In [14]:
# accessing first row of the dataframe
store_data[:1]

CustomerID location gender         type  quantity  total_bill
0   CustID00  Chicago      M  Electronics         1         100

In [15]:
# accessing first column of the dataframe
store_data['location']

0          Chicago
1           Boston
2          Seattle
3    San Francisco
4           Austin
Name: location, dtype: object

In [16]:
# accessing rows with the step size of 2
store_data[::2]

CustomerID location gender            type  quantity  total_bill
0   CustID00  Chicago      M     Electronics         1         100
2   CustID02  Seattle      F  Food&Beverages         4         125
4   CustID04   Austin      F          Beauty         1          80

In [17]:
# accessing the rows in reverse
store_data[::-2]

CustomerID location gender            type  quantity  total_bill
4   CustID04   Austin      F          Beauty         1          80
2   CustID02  Seattle      F  Food&Beverages         4         125
0   CustID00  Chicago      M     Electronics         1         100

**Using loc and iloc method**

**loc method**

* loc is a  method to access rows and columns on pandas objects. When using the loc method on a dataframe, we specify which rows and which columns we want by using the following format:

  * **dataframe.loc[row selection, column selection]**

* DataFrame.loc[] method is a method that takes **only index labels** and returns row or dataframe if the index label exists in the data frame.

In [18]:
# accessing first index value using loc method (indexing starts from 0 in python)
store_data.loc[1]

CustomerID          CustID01
location              Boston
gender                     M
type          Food&Beverages
quantity                   3
total_bill                75
Name: 1, dtype: object

**Accessing selected rows and columns using loc method**

In [19]:
# accessing 1st and 4th index values along with location and type columns
store_data.loc[[1,4],['location','type']]

location            type
1   Boston  Food&Beverages
4   Austin          Beauty

**iloc method**

* The iloc indexer for Pandas Dataframe is used for **integer location-based** indexing/selection by position. When using the loc method on a dataframe, we specify which rows and which columns we want by using the following format:

  * **dataframe.iloc[row selection, column selection]**



In [20]:
# accessing selected rows and columns using iloc method
store_data.iloc[[1,4],[0,2]]

CustomerID gender
1   CustID01      M
4   CustID04      F

**Difference between loc and iloc indexing methods**

* loc is label-based, which means that you have to specify rows and columns based on their row and column labels.
* iloc is integer position-based, so you have to specify rows and columns by their integer position values (0-based integer position).


If we use labels instead of index values in .iloc it will throw an error.

In [21]:
# accessing selected rows and columns using iloc method
store_data.iloc[[1,4],['location','type']]

IndexError: .iloc requires numeric indexers, got ['location' 'type']

* As expected, .iloc has given error on using 'labels'.

We can modify entries of a dataframe using loc or iloc too

In [22]:
print(store_data.loc[4,'type'])
store_data.loc[4,'type'] = 'Electronics'

Beauty


In [23]:
store_data

CustomerID       location gender            type  quantity  total_bill
0   CustID00        Chicago      M     Electronics         1         100
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50
4   CustID04         Austin      F     Electronics         1          80

In [24]:
store_data.iloc[4,3] = 'Beauty'
store_data

CustomerID       location gender            type  quantity  total_bill
0   CustID00        Chicago      M     Electronics         1         100
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50
4   CustID04         Austin      F          Beauty         1          80

### 1.15 Pandas - Saving and Loading DataFrames

**Note**

In real-life scenario, we deal with much larger datasets that have thousands of rows and multiple columns. It will not be feasible for us to create datasets using multiple lists, especially if the number of columns and rows increases.

So, it is clear we need a more efficient way of handling the data simultaneously at the columns and row levels. In Python, we can import dataset from our local system, from links, or from databases and work on them directly instead of creating our own dataset.

**Loading a CSV file in Python**

**For Jupyter Notebook**
* When the data file and jupyter notebook are in the same folder.

In [ ]:
# Using pd.read_csv() function will work without any path if the notebook and dataset are in the folder

# data = pd.read_csv('StockData.csv')

**For Google Colab with Google Drive**

First, we have to give google colab access to our google drive:

In [28]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Once we have access we can load files from google drive using read_csv() function.

In [34]:
path="/content/drive/MyDrive/GenAI/StockData.csv"
data=pd.read_csv(path)

In [35]:
# head() function helps us to see the first 5 rows of the data
data.head()

stock        date    price
0  AAPL  08-02-2013  67.8542
1  AAPL  11-02-2013  68.5614
2  AAPL  12-02-2013  66.8428
3  AAPL  13-02-2013  66.7156
4  AAPL  14-02-2013  66.6556

**Loading an excel file in Python**

In [36]:
path_excel="/content/drive/MyDrive/GenAI/StockData.xlsx"
data_excel = pd.read_excel(path_excel)

In [37]:
data_excel.head()

stock       date    price
0  AAPL 2013-02-08  67.8542
1  AAPL 2013-02-11  68.5614
2  AAPL 2013-02-12  66.8428
3  AAPL 2013-02-13  66.7156
4  AAPL 2013-02-14  66.6556

**Saving a dataset in Python**

**Saving the dataset as a csv file**

To save a dataset as .csv file the syntax used is -

**data.to_csv('name of the file.csv', index=False)**

In [38]:
data.to_csv('/content/drive/MyDrive/GenAI/Saved_StockData.csv',index=False)

* In jupyter notebook, the dataset will be saved in the folder where the jupyter notebook is located.
* We can also save the dataset to a desired folder by providing the path/location of the folder.

**Saving the dataset as an excel spreadsheet**

To save a dataset as .xlsx file the syntax used is -

**data.to_excel('name of the file.xlsx',index=False)**

In [39]:
data.to_excel('/content/drive/MyDrive/GenAI/Saved_StockData.xlsx',index=False)

### 1.16 Pandas - Functions

**head() - to check the first 5 rows of the dataset**

In [40]:
data.head()

stock        date    price
0  AAPL  08-02-2013  67.8542
1  AAPL  11-02-2013  68.5614
2  AAPL  12-02-2013  66.8428
3  AAPL  13-02-2013  66.7156
4  AAPL  14-02-2013  66.6556

**tail() - to check the last 5 rows of the dataset**

In [41]:
data.tail()

stock        date  price
5031   ZTS  01-02-2018  77.82
5032   ZTS  02-02-2018  76.78
5033   ZTS  05-02-2018  73.83
5034   ZTS  06-02-2018  73.27
5035   ZTS  07-02-2018  73.86

**shape - to check the number of rows and columns in the dataset**

In [42]:
data.shape

(5036, 3)

In [43]:
data.size

15108

In [46]:
data.columns

Index(['stock', 'date', 'price'], dtype='object')

In [47]:
data.values

array([['AAPL', '08-02-2013', 67.8542],
       ['AAPL', '11-02-2013', 68.5614],
       ['AAPL', '12-02-2013', 66.8428],
       ...,
       ['ZTS', '05-02-2018', 73.83],
       ['ZTS', '06-02-2018', 73.27],
       ['ZTS', '07-02-2018', 73.86]], dtype=object)

In [57]:
data.ndim

2

In [52]:
data.columns

Index(['stock', 'date', 'price'], dtype='object')

In [58]:
data.dtypes

stock     object
date      object
price    float64
dtype: object

* The dataset has 5036 rows and 3 columns.

**info() - to check the data type of the columns**

In [59]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5036 entries, 0 to 5035
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   stock   5036 non-null   object 
 1   date    5036 non-null   object 
 2   price   5036 non-null   float64
dtypes: float64(1), object(2)
memory usage: 118.2+ KB


* The price column is numeric in nature while the stock and date columns are of object types.

**min() - to check the minimum value of a numeric column**

In [61]:
data['price'].min()

28.4

**max() - to check the maximum value of a numeric column**

In [62]:
data['price'].max()

179.26

**unique() - to check the number of unique values that are present in a column**

In [63]:
data['stock'].unique()

array(['AAPL', 'SNI', 'TJX', 'ZTS'], dtype=object)

**value_counts() - to check the number of values that each unique quantity has in a column**

In [64]:
data['stock'].value_counts()

stock
AAPL    1259
SNI     1259
TJX     1259
ZTS     1259
Name: count, dtype: int64

**value_counts(normalize=True) - using the `normalize` parameter and initializing it to True will return the relative frequencies of the unique values.**

In [65]:
data['stock'].value_counts(normalize=True)

stock
AAPL    0.25
SNI     0.25
TJX     0.25
ZTS     0.25
Name: proportion, dtype: float64

**Statistical Functions**

**mean() - to check the mean (average) value of the column**

In [66]:
data['price'].mean()

np.float64(73.05702966640192)

**median() - to check the median value of the column**

In [67]:
data['price'].median()

69.08500000000001

**mode() - to check the mode value of the column**

In [69]:
data['stock'].mode()

0    AAPL
1     SNI
2     TJX
3     ZTS
Name: stock, dtype: object

**To access a particular mode when the dataset has more than 1 mode**

In [72]:
#to access the first mode
data['price'].mode()[0]

np.float64(74.59)